consensus/src/quorum_store/batch_aggregator.rs

```diff
@@ -0,0 +1,227 @@
+                    "No fragments in batch aggregator state"
+                );
+
+                // In this case, the next fragment must start self_batch_id + 1
+                batch_id <= self_batch_id
+            } else {
+                (batch_id, fragment_id) < (self_batch_id, self.next_fragment_id())
+            }
+        } else {
+            false
+        }
+    }
+
+    // Should only be called with a non-outdated fragment.
+    fn missed_fragment(&self, batch_id: BatchId, fragment_id: usize) -> bool {
+        match self.batch_id {
+            Some(self_batch_id) => {
+                if batch_id > self_batch_id {
+                    self.batch_state.is_some() || fragment_id > 0
+                } else {
+                    assert!(
+                        batch_id == self_batch_id,
+                        "Missed fragment called with an outdated fragment"
+                    );
+                    fragment_id > self.next_fragment_id()
+                }
+            }
+            // Allow larger batch_id (> 0) the first time as quorum store might
+            // be recovering from a crash and continuing with a larger batch_id.
+            None => fragment_id > 0,
+        }
+    }
+
+    /// Appends transactions from a batch fragment, ensuring that the fragment is
+    /// consistent with the state being aggregated, and the maximum batch size is
+    /// being respected. Otherwise, a corresponding error is returned which should
+    /// be handled on the caller side (i.e. panic for self, log & ignore for peers).
+    pub(crate) fn append_transactions(
+        &mut self,
+        batch_id: BatchId,
+        fragment_id: usize,
+        transactions: Vec<SerializedTransaction>,
+    ) -> Result<(), BatchAggregatorError> {
+        if self.outdated_fragment(batch_id, fragment_id) {
+            counters::EXPIRED_BATCH_FRAGMENTS_COUNT.inc();
+            // Replay or batch / fragment received out of order.
+            return Err(BatchAggregatorError::OutdatedFragment);
+        }
+
+        let missed_fragment = self.missed_fragment(batch_id, fragment_id);
+        if missed_fragment {
+            counters::MISSED_BATCH_FRAGMENTS_COUNT.inc();
+            debug!(
+                "QS: missed_fragment batch_id: {:?} fragment_id {:?}",
+                batch_id, fragment_id
+            );
+            // If we started receiving a new batch, allow aggregating it by
+            // clearing the state. Otherwise, when some fragment is skipped
+            // within a batch, this just stops aggregating it.
+            self.batch_state = None;
+        }
+
+        // Fragment wasn't outdated, fast forward batch id.
+        self.batch_id = Some(batch_id);
+        if fragment_id == 0 {
+            // If a fragment was missed, state should be cleared
+            // above, and otherwise, it must be cleared by end_batch.
+            debug_assert!(self.batch_state.is_none(), "Batch state not cleared");
```

Comment:
let's not use debug_assert, or at least only debug assert
i.e. add error!() log here

in general, with debug_assert, you cannot really test that the line below it doesn't have a bug.

Reply:
should be done